File: python/rectangles/rectangles.py
"""
For a given diagram calculate the number of rectangles it contains
"""

HORIZONTAL = '-+'
VERTICAL = '|+'
CORNER = '+'

def rectangles(strings):
    """
    Count the number of rectangles in the given diagram.
    :param strings: List of strings, should all be the same length
    :returns: count of rectangles found
    """
    # turn strings into a list of list of chars for easier indexing
    ret = 0
    board = [list(row) for row in strings]
    for row_index, row in enumerate(board):
        for col_index, rune in enumerate(row):
            if rune in CORNER:
                ret += trace_rectangle(col_index, row_index, board)
    return ret

def trace_rectangle(col_index, row_index, board):
    """
    See if you can trace out a rectangle from this location. Try to 
    keep going and branch out when we encounter a corner character.
    :param col_index: column index to start at
    :param row_index: row index to start at
    :returns: number of rectangles traced out starting at this position.
    """
    col_count = len(board[0])

    ret = 0

    # trace left
    current_x = col_index + 1
    current_y = row_index
    while current_x < col_count and board[current_y][current_x] in HORIZONTAL:
        # if corner try to turn there.
        if board[current_y][current_x] == CORNER:
            ret += trace_down(col_index, row_index, current_x, current_y, board)
        current_x += 1
    return ret 

def trace_down(left, top, col_index, row_index, board):
    """
    Search downard for a corner to branch on or non-vertial to stop at.
    :param left: column where the rectangle started
    :param top: row where the rectangle started
    :param col_index: column index to start at - top right corner
    :param row_index: row index to start at - top right corner
    :returns: number of rectangles traced out.
    """
    
    row_count = len(board)

    ret = 0
    # trace down
    current_x = col_index
    current_y = row_index + 1
    while current_y < row_count and board[current_y][current_x] in VERTICAL:
        # if corner try to turn there.
        if board[current_y][current_x] == CORNER:
            ret += trace_right(left, top, current_x, current_y, board)
        current_y += 1
    return ret 

def trace_right(left, top, right, bottom, board):
    """
    Search right to see if we have a corner at the expected spot and horizontal characters inbetween
    No rectangles if we encounter an unexpected character
    :param left: column where the rectangle started
    :param top: row where the rectangle started
    :param right: column where the right side of the rectangle should be
    :param bottom: row where the bottom of the rectangle should be
    :returns: number of rectangles traced out.
    """

    current_x = right
    current_y = bottom

    # make sure the corner is where it should be
    if board[bottom][left] != CORNER:
        return 0
    
    while current_x >= left and board[current_y][current_x] in HORIZONTAL:
        current_x -= 1
    
    # did we stop early
    if current_x != left - 1:
        return 0
    
    return trace_up(left, top, bottom, board)

def trace_up(left, top, bottom, board):
    """
    Search up to see if we have a corner at the expected spot and vertical characters inbetween
    No rectangles if we encounter an unexpected character
    :param left: column where the rectangle started
    :param top: row where the rectangle started
    :param bottom: row where the bottom of the rectangle should be
    :returns: number of rectangles traced out.
    """

    current_x = left
    current_y = bottom

    # make sure the corner is where it should be
    if board[top][left] != CORNER:
        return 0
    
    while current_y >= top and board[current_y][current_x] in VERTICAL:
        current_y -= 1
    
    # did we stop early
    if current_y != top - 1:
        return 0
    
    # one rectangle found    
    return 1

File: go/rectangles/rectangles.go
// Rectangles exercise - count the rectangles on a grid.
package rectangles

// Find the number of rectangles on the given diagram.
// @param diagram: slice of strings that rectangles are plotted on.
// @returns: Count of rectangles found.
func Count(diagram []string) int {
	var rows = len(diagram)
	var cols = 0
	data := make([][]rune, rows)
	for i, row := range diagram {
		data[i] = []rune(row)
		if len(row) > cols {
			cols = len(row)
		}
	}
	var countRectangles int = 0

	if rows == 0 || cols == 0 {
		return 0
	}

	for y, row := range data {
		for x, char := range row {
			if char == '+' {
				countRectangles += Probe(data, cols, rows, x, y)
			}
		}
	}

	return countRectangles
}

// Look for retangles with the top left corner at x, y
// @param data: The 2d slice of runes with the grid to find rectangles on
// @param cols: Number of columns in the data grid
// @param rows: Number of rows in the data grid
// @param x: x-coordinate of the top left corner
// @param y: y-coordinate of the top left corner
// @returns: count of rectangles found with a top left corner at x, y.
func Probe(data [][]rune, cols, rows, x, y int) int {
	// Probe left
	var x1 int = x + 1
	var y1 int = y
	var countRectangles int = 0

	for x1 < cols {
		var char rune = data[y1][x1]
		if char == ' ' {
			return countRectangles
		}
		if char == '+' {
			countRectangles += ProbeDown(data, cols, rows, x, y, x1)
		}
		if char != '-' && char != '+' {
			return countRectangles
		}
		x1++
	}
	return countRectangles
}

// Search for the bottom right corner of a rectangle.
// @param data: The 2d slice of runes with the grid to find rectangles on
// @param cols: Number of columns in the data grid
// @param rows: Number of rows in the data grid
// @param x: x-coordinate of the top left corner
// @param y: y-coordinate of the top left corner
// @param x1: x-coordinate of the top right corner
// @returns: count of rectangles found with (x,y)-(x1,?)
func ProbeDown(data [][]rune, cols, rows, x, y, x1 int) int {
	var y1 int = y + 1
	var countRectangles int = 0

	for y1 < rows {
		var char rune = data[y1][x1]
		if char == ' ' {
			return countRectangles
		}
		if char == '+' {
			countRectangles += ValidateRect(data, x, y, x1, y1)
		}
		if char != '|' && char != '+' {
			return countRectangles
		}
		y1++
	}
	return countRectangles
}

// Ensure that a pair of x, y coordinates contain a rectangle.
// @param data: The 2d slice of runes with the grid to find rectangles on
// @param x1: x-coordinate of the top left corner
// @param y1: y-coordinate of the top left corner
// @param x2: x-coordinate of the bottom right corner
// @param y2: y-coordinate of the bottom right corner
// @returns: 1 if rectangle found, and 0 if it was not.
func ValidateRect(data [][]rune, x1, y1, x2, y2 int) int {
	// Check the corners
	if data[y1][x1] != '+' || data[y1][x2] != '+' || data[y2][x1] != '+' || data[y2][x2] != '+' {
		return 0
	}

	// Check top/bottom
	for x := x1 + 1; x < x2; x++ {
		if data[y1][x] != '+' && data[y1][x] != '-' {
			return 0
		}
		if data[y2][x] != '+' && data[y2][x] != '-' {
			return 0
		}
	}

	// Check left/right
	for y := y1 + 1; y < y2; y++ {
		if data[y][x1] != '+' && data[y][x1] != '|' {
			return 0
		}
		if data[y][x2] != '+' && data[y][x2] != '|' {
			return 0
		}
	}
	return 1
}
